satellite/satellitedb/orders.go

```diff
@@ -214,3 +218,191 @@ func (db *ordersDB) UnuseSerialNumber(ctx context.Context, serialNumber storj.Se
+		if err == nil {
+			err = tx.Commit()
+		} else {
+			err = errs.Combine(err, tx.Rollback())
+		}
+	}()
+
+	now := time.Now().UTC()
+	intervalStart := time.Date(now.Year(), now.Month(), now.Day(), now.Hour(), 0, 0, 0, now.Location())
+
+	rejectedRequests := make(map[storj.SerialNumber]bool)
+	reject := func(serialNumber storj.SerialNumber) {
+		r := &pb.SettlementResponse{
+			SerialNumber: serialNumber,
+			Status:       pb.SettlementResponse_REJECTED,
+		}
+		rejectedRequests[serialNumber] = true
+		responses = append(responses, r)
+	}
+
+	// processes the insert to used serials table individually so we can handle
+	// the case where the order has already been process.  Duplicates and previously
+	// processed orders are rejected
+	for _, request := range requests {
+		stmt, err := db.buildInsertUseSerialStatement(request.OrderLimit)
+		if err != nil {
+			return nil, err
+		}
+		_, err = tx.Exec(stmt)
+		if err != nil {
+			if pgutil.IsConstraintError(err) || sqliteutil.IsConstraintError(err) {
+				reject(request.OrderLimit.SerialNumber)
+			} else {
+				return nil, Error.Wrap(err)
+			}
+		}
+	}
+
+	// call to get all the bucket IDs
+	query := db.buildGetBucketIdsQuery(len(requests))
+	statement := db.db.Rebind(query)
+
+	args := make([]interface{}, 0, len(requests))
+	for _, request := range requests {
+		args = append(args, request.OrderLimit.SerialNumber.Bytes())
+	}
+
+	rows, err := tx.Query(statement, args...)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	bucketMap := make(map[storj.SerialNumber][]byte)
+	for rows.Next() {
+		var serialNumber, bucketID []byte
+		err := rows.Scan(&serialNumber, &bucketID)
+		if err != nil {
+			return nil, errs.Wrap(err)
+		}
+		sn, err := storj.SerialNumberFromBytes(serialNumber)
+		if err != nil {
+			return nil, errs.Wrap(err)
+		}
+		bucketMap[sn] = bucketID
+	}
+
+	// build all the bandwidth updates into one sql statement
+	var updateRollupStatement string
+	for _, request := range requests {
+		_, rejected := rejectedRequests[request.OrderLimit.SerialNumber]
+		if !rejected {
+			bucketID, ok := bucketMap[request.OrderLimit.SerialNumber]
+			if ok {
+				projectID, bucketName, err := orders.SplitBucketID(bucketID)
+				if err != nil {
+					return nil, errs.Wrap(err)
+				}
+
+				stmt, err := db.buildUpdateBucketBandwidthRollupStatements(request.OrderLimit, request.Order, projectID[:], bucketName, intervalStart)
+				if err != nil {
+					return nil, err
+				}
+				updateRollupStatement += stmt
+
+				stmt, err = db.buildUpdateStorageNodeBandwidthRollupStatements(request.OrderLimit, request.Order, intervalStart)
+				if err != nil {
+					return nil, err
+				}
+				updateRollupStatement += stmt
+			} else {
+				reject(request.OrderLimit.SerialNumber)
+			}
+		}
+	}
+
+	_, err = tx.Exec(updateRollupStatement)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	for _, request := range requests {
+		_, rejected := rejectedRequests[request.OrderLimit.SerialNumber]
+		if !rejected {
```

Comment:
Similar stylistic comment here, I really like:

```
if rejected {
    continue
}
```

then you can do without the extra indentation and closure.

Reply:
Updated

Reply:
I think you missed this one. Otherwise, I'm approving.